Introduce RefM, which is like Ref but can handle effectful atomic updates

jdegoes opened the pull request:
I've found myself reinventing this over many different projects. Better to live in ZIO.

It's a version of `Ref` whose `update` and `modify` methods accept effectful update/modify functions.

The API is otherwise identical. I also pared down the `Ref` API, deleting useless methods.

wi101 commented on core/shared/src/main/scala/scalaz/zio/Ref.scala:
```diff
@@ -35,12 +35,6 @@ final class Ref[A] private (private val value: AtomicReference[A]) extends AnyVa
    */
   final def setLater(a: A): IO[Nothing, Unit] = IO.sync(value.lazySet(a))
 
-  /**
```
Update the test cases in `RefSpec` where we used `setLater` & `trySet` & `compareAndSet`

wi101 commented on core/shared/src/main/scala/scalaz/zio/RefM.scala:
```diff
@@ -0,0 +1,87 @@
+ * } yield ()
+ * }}}
+ */
+final class RefM[A] private (private val value: Ref[A], private val queue: Queue[RefM.Bundle[A, _]])
```
I suggest if we define value and queue without `val` they would be inaccessible :
 ```scala
final class RefM[A] private (value: Ref[A], queue: Queue[RefM.Bundle[A, _]])
```

wi101 commented on core/shared/src/main/scala/scalaz/zio/RefM.scala:
```diff
@@ -0,0 +1,87 @@
+
+object RefM extends Serializable {
+  private[RefM] final case class Bundle[A, B](
+    var interrupted: Option[List[Throwable]] = None,
```
could we use `Ref` instead of `var`?

wi101 commented on core/shared/src/main/scala/scalaz/zio/RefM.scala:
```diff
@@ -0,0 +1,87 @@
+    final def run(a: A): IO[List[Throwable], A] =
+      interrupted match {
+        case Some(ts) => IO.fail(ts)
+        case None     => update(a).flatMap(t => promise.complete(t._1).map(_ => t._2))
```
I prefer this ^^: 
```scala
.flatMap{ case (b, a) => promise.complete(b).map(_ => a) }
```

regiskuckaertz replied:
Even better: `map(_ => a)` → `const(a)`

regiskuckaertz commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -338,6 +338,20 @@ sealed abstract class IO[+E, +A] { self =>
+    IO.bracket0(IO.unit)(
+      (_, eb: ExitResult[E, A]) =>
+        eb match {
+          case ExitResult.Completed(_)   => IO.unit
```
These two cases could be merged in a single fallback case with no pattern match

regiskuckaertz commented on core/shared/src/main/scala/scalaz/zio/RefM.scala:
```diff
@@ -0,0 +1,87 @@
+    for {
+      ref   <- Ref(a)
+      queue <- Queue.bounded[Bundle[A, _]](n)
+      _     <- queue.take.flatMap(b => ref.get.flatMap(a => b.run(a).redeem(onDefect, ref.set))).forever.fork
```
Beautiful! So `Bundle` allows you to run all the updates in a separate fiber?

I thought of another option you might want to add: updates could be composed in a `KleisliIO` and run only when someone calls `get`.

jdegoes commented:
@wi101 @regiskuckaertz Addressed all feedback! Ready for another look if you have time.

@regiskuckaertz The only reason I didn't implement `Ref` in terms of `RefM` is that `Ref` has far better performance. I think specializing that on atomic compare-and-swap provided by `AtomicReference` makes sense purely for performance reasons.

wi101 commented:
@jdegoes if `RefM` is more performant, can we replace Ref with it? (while it has the same methods, it wouldn't affect to the existed code for example in Queue) 
And we use `AtomicReference` in `RefM` instead the old `Ref`, could we do that?
 What do you think?

jdegoes commented:
@wi101 I mean that `Ref` is much faster than `RefM`. But on the other hand, with `RefM`, you can perform effects inside the `update` / `modify` methods. So `RefM` is much more powerful. `RefM` is implemented with the help of `Ref`, `Promise`, and `Queue`.

wi101 commented:
@jdegoes Sorry, I didn't read the comment above -_- and I was totally wrong.. RefM already use Queue -_- 

It makes sense, Thanks.
